### PR TITLE
Implement GetLeavesByIndex in server

### PR DESCRIPTION
**Pull request by Martin2112**
Also fix up some errors with pointer types, rename server go file to correct style, add tests and stuff, add options to client for what leaves get requested.

### CODE REVIEW

File: server/trillian_log_server.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+func (t *TrillianLogServer) GetLatestSignedLogRoot(ctx context.Context, req *trillian.GetLatestSignedLogRootRequest) (*trillian.GetLatestSignedLogRootResponse, error) {
+	return nil, ErrNotImplemented
+}
+
+func (t *TrillianLogServer) GetSequencedLeafCount(ctx context.Context, req *trillian.GetSequencedLeafCountRequest) (*trillian.GetSequencedLeafCountResponse, error) {
+	return nil, ErrNotImplemented
+}
+
+func (t *TrillianLogServer) GetLeavesByIndex(ctx context.Context, req *trillian.GetLeavesByIndexRequest) (*trillian.GetLeavesByIndexResponse, error) {
```
Comment (AlCutter) [inline server/trillian_log_server.go]: godoc missing on the exported funcs

File: server/trillian_log_server.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+	if err != nil {
+		return nil, err
+	}
+
+	if !validateLeafIndices(req.LeafIndex) {
+		return &trillian.GetLeavesByIndexResponse{Status: buildStatusWithDesc(trillian.TrillianApiStatusCode_ERROR, "Invalid -ve leaf index in request")}, nil
+	}
+
+	// Read only transaction, failure to commit not propagated to client
```
Comment (AlCutter) [inline server/trillian_log_server.go]: Don't think you can do that - if the commit fails then the data read under the TX is invalid...

File: server/trillian_log_server.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+func (t *TrillianLogServer) GetLeavesByIndex(ctx context.Context, req *trillian.GetLeavesByIndexRequest) (*trillian.GetLeavesByIndexResponse, error) {
+	tx, err := t.prepareStorageTx(*req.LogId)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if !validateLeafIndices(req.LeafIndex) {
```
Comment (AlCutter) [inline server/trillian_log_server.go]: worth fetching the latest root and ensuring the req'd leaves are within the tree size?

File: server/trillian_log_server.go (hunk 2 above)
Comment (Martin2112) [inline server/trillian_log_server.go]: That would appear to be a flaw in the cunning plan. I'll fix it in next push.

File: server/trillian_log_server.go (hunk 3 above)
Comment (Martin2112) [inline server/trillian_log_server.go]: Wasn't sure about the overhead of that if someone is trying to read the whole log in batches. Not clear to me where this state should be kept and if it should be cached. Can I todo it for now?

File: server/trillian_log_server.go (hunk 1 above)
Comment (Martin2112) [inline server/trillian_log_server.go]: Will fix.

File: server/trillian_log_server.go [hunk 4]
```diff
@@ -75,26 +75,30 @@ func (t *TrillianLogServer) GetSequencedLeafCount(ctx context.Context, req *tril
...
 
 	if err != nil {
+		tx.Rollback()
 		return nil, err
 	}
 
 	leafProtos := leavesToProtos(leaves)
+
+	if err := t.commitAndLog(tx, "GetLeavesByIndex"); err != nil {
+		tx.Rollback()
```
Comment (AlCutter) [inline server/trillian_log_server.go]: Is it ok to call `Rollback()` if you already called `Commit()` and it failed?

Comment (Martin2112) [inline server/trillian_log_server.go]: It's doubtful but this is an application level tx object. There could be stuff that isn't cleaned up that we own, though there currently isn't.

Comment (Martin2112) [inline server/trillian_log_server.go]: Actually we could define the API so the TX impl must clean up if the commit fails then not worry about it. Is that better?

Comment (AlCutter) [inline server/trillian_log_server.go]: Maybe, dunno yet :)

**Comment by Martin2112** [comment]
Implemented the change so it no longer calls rollback if commit fails.

**Pull request merged**